test/functional/legacy/080_substitute_spec.lua

```diff
@@ -0,0 +1,177 @@
+-- Test for *sub-replace-special* and *sub-replace-expression* on substitue().
+-- Test for submatch() on substitue().
+-- Test for *:s%* on :substitute.
+
+local helpers = require('test.functional.helpers')
+local feed, insert, source = helpers.feed, helpers.insert, helpers.source
+local clear, execute, expect = helpers.clear, helpers.execute, helpers.expect
+local eq, eval = helpers.eq, helpers.eval
+
+describe('substitue()', function()
+  before_each(clear)
+
+  local function test_1_and_2()
+    eq('AA', eval("substitute('A', 'A', '&&', '')"))
+    eq('&', eval([[substitute('B', 'B', '\&', '')]]))
+    eq('C123456789987654321', eval([[substitute('C123456789', ]] ..
+      [['C\(.\)\(.\)\(.\)\(.\)\(.\)\(.\)\(.\)\(.\)\(.\)', ]] ..
+      [['\0\9\8\7\6\5\4\3\2\1', '')]]))
+    eq('d', eval("substitute('D', 'D', 'd', '')"))
+    eq('~', eval("substitute('E', 'E', '~', '')"))
+    eq('~', eval([[substitute('F', 'F', '\~', '')]]))
+    eq('Gg', eval([[substitute('G', 'G', '\ugg', '')]]))
+    eq('Hh', eval([[substitute('H', 'H', '\Uh\Eh', '')]]))
+    eq('iI', eval([[substitute('I', 'I', '\lII', '')]]))
+    eq('jJ', eval([[substitute('J', 'J', '\LJ\EJ', '')]]))
+    eq('Kk', eval([[substitute('K', 'K', '\Uk\ek', '')]]))
+    eq('l\rl', eval("substitute('lLl', 'L', '\r', '')"))
+    eq('m\rm', eval([[substitute('mMm', 'M', '\r', '')]]))
+    eq('n\rn', eval("substitute('nNn', 'N', '\\\r', '')"))
+    eq('o\no', eval([[substitute('oOo', 'O', '\n', '')]]))
+    eq('p\bp', eval([[substitute('pPp', 'P', '\b', '')]]))
+    eq('q\tq', eval([[substitute('qQq', 'Q', '\t', '')]]))
+    eq('r\\r', eval([[substitute('rRr', 'R', '\\', '')]]))
+    eq('scs', eval([[substitute('sSs', 'S', '\c', '')]]))
+    -- There is no line with t's in the original legacy test part 1.  The same
+    -- test was done with x's.  Part two did the following test with t's and
+    -- moved the test here done with y and z up to x and y.
+    eq('t\rt', eval([[substitute('tTt', 'T', "\r", '')]]))
+    eq('u\nu', eval([[substitute('uUu', 'U', "\n", '')]]))
+    eq('v\bv', eval([[substitute('vVv', 'V', "\b", '')]]))
+    eq('w\\w', eval([[substitute('wWw', 'W', "\\", '')]]))
+    -- See the line with t's above.
+    eq('YyyY', eval([[substitute('Y', 'Y', '\L\uyYy\l\EY', '')]]))
+    eq('zZZz', eval([[substitute('Z', 'Z', '\U\lZzZ\u\Ez', '')]]))
+  end
+
+  it('with "set magic" (TEST_1)', function()
```

Comment:
these where the "headers" in the original test for the different parts. To say it differently: The original test was devided by TEST_X strings and I used these places to split it into seperated it() blocks. I left the strings there to make it easier to incorporate changes from upstream.